libc/nolibc: Implement LWIP-compatible interface (given Musl)

razvand opened the pull request:
<!--

Thank you for opening a new PR to the Unikraft Open Source Project!  We welcome
new changes, features, fixes, and more!  Please fill in this form to indicate
the status of your PR.  Please ensure you have read the contribution guidelines
before opening a new PR as this will cover the PR process:

  https://unikraft.org/docs/contributing/

-->

### Prerequisite checklist

<!--
Please mark items appropriately:
-->

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [ ] Ran the [`checkpatch.pl`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.pl) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]

### Description of changes

Move items from Musl and LWIP to nolibc. This way LWIP is compatible both with Musl and with nolibc. The [Musl-support changes](https://github.com/unikraft/lib-lwip/pull/23) must be applied to LWIP.

craciunoiuc reviewed (commented):
I'm a bit lost in this 115 commit PR. I only looked on the last commit as it seems related to the PR title.

I see plenty of commented lines so I suppose this is still in the early stages. I will look over it once it's ready/more ready.

skuenzer commented:
> 

Knowing what @razvand did :wink:, it is baed on draft PRs of the scheduling API and others. The relevant commits of this PR are just the recent two from @razvand. He will be able to rebase his commits after we could integrate scheduling and clone system call patches.

razvand committed:
lib/nolibc: Use sys/socket.h from Musl

LWIP has been updated to work with Musl. Headers and source code files
from LWIP have been removed since they are part of Musl.

In order to keep nolibc working with LWIP, previous items part of LWIP
now have to be part of nolibc.

This commit removes the current version of sys/socket.h and replaces it
with a fuller version from Musl.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
lib/nolibc: Add IOV_MAX macro (required by LWIP)

LWIP has been updated to work with Musl. Headers and source code files
from LWIP have been removed since they are part of Musl.

In order to keep nolibc working with LWIP, previous items part of LWIP
now have to be part of nolibc.

This commit adds the IOV_MAX macro in nolibc. The macro was previously
part of LWIP but has now been removed to make it compatible with Musl.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
lib/nolibc: Add ntohs and htons

LWIP has been updated to work with Musl. Headers and source code
files from LWIP have been removed since they are part of Musl.

In order to keep nolibc working with LWIP, previous items part of
LWIP now have to be part of nolibc.

This commit adds the implementation of `htons()` and `ntohs()`,
functions that had been previously defined in LWIP, but have now been
removed to make it compatible with Musl.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
lib/nolibc: Add definition of h_errno

LWIP has been updated to work with Musl. Headers and source code files
from LWIP have been removed since they are part of Musl.

In order to keep nolibc working with LWIP, previous items part of LWIP
now have to be part of nolibc.

This commit adds the definition of the `h_errno` variable.  It had been
previously defined in LWIP, but has now been removed to make it
compatible with Musl.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
lib/nolibc: Add net/if.h from Musl

LWIP has been updated to work with Musl. Headers and source code files
from LWIP have been removed since they are part of Musl.

In order to keep nolibc working with LWIP, previous items part of LWIP
now have to be part of nolibc.

This commit adds the net/if.h header from Musl. It was previously part
of LWIP, but has now been removed to make it compatible with Musl.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

jschlumpp commented on lib/nolibc/musl-imported/include/sys/socket.h:
```diff
@@ -0,0 +1,378 @@
+#include <nolibc-internal/shareddefs.h>
+
+/*
+ * FIXME: These includes shouldn't be here.
```
Wouldn't be a patch for LWIP more appropriate?

razvand replied:
Yes, I will do a PR for LWIP and move it there.

jschlumpp commented on lib/nolibc/musl-imported/include/sys/socket.h:
```diff
@@ -0,0 +1,378 @@
+#include <fcntl.h>
+#include <sys/ioctl.h>
+
+#ifdef CONFIG_LIBPOSIX_SOCKET
```
Is this necessary for software that doesn't need sockets but still has a `#include <sys/socket.h>` somewhere? Otherwise I would just remove the `#ifdef`.

razvand replied:
I removed it.

craciunoiuc commented on lib/nolibc/musl-imported/include/sys/socket.h:
```diff
@@ -0,0 +1,378 @@
+
+#ifdef __cplusplus
+}
+#endif
+#endif
```
Add the trailing comment to all `#endif`s in this commit.

craciunoiuc replied:
Not sure if this is just imported, or changed a bit.

If it's just imported, then no need to do this.

razvand replied:
Yes, it's imported from Musl. I will leave it like this.

craciunoiuc commented on lib/nolibc/musl-imported/src/network/h_errno.c:
```diff
@@ -0,0 +1,9 @@
+#include <netdb.h>
+
+#undef h_errno
```
Is this actually needed? The header above has no `h_errno` define so this will not do anything.

craciunoiuc reviewed (commented):
Some small remarks from my side, will come back after it's mergeable.

jschlumpp commented on lib/nolibc/include/netdb.h:
```diff
@@ -0,0 +1,14 @@
+extern "C" {
+#endif
+
+extern int h_errno;
```
Usually this is a macro for `*__h_errno_location()` that returns the correct value and that function is also defined in the other file. But it's not used for some reason. 
Did you forget to do it that way or is this intentional?

jschlumpp commented on lib/nolibc/musl-imported/src/network/h_errno.c:
```diff
@@ -0,0 +1,9 @@
+#undef h_errno
+int h_errno;
+
+int *__h_errno_location(void)
```
This is currently neither exported nor used (see other comment for `netdb.h`

jschlumpp reviewed (changes requested):
I only looked at the latest five commits (I guess the other commits are only included for context)

StefanJum commented:
@razvand everything seems to work fine, after you rebase it and address @mschlumpp and @craciunoiuc s comments I will add my tag.

Also I think the [last commit message](https://github.com/unikraft/unikraft/commit/3f92b632baad25803b8c110aa44206ef5466bc9b) should start with `lib/nolibc:`, not just `nolibc:`, just to be consistent with the other 4 commits.

razvand commented:
@mschlumpp , @craciunoiuc , @StefanJum , please take another look in this PR.

[LWIP PR 27](https://github.com/unikraft/lib-lwip/pull/27) needs to be included as well for this to work.

StefanJum commented:
@razvand this needs rebasing, there are conflicts with the staging branch.

craciunoiuc reviewed (commented):
I'm okay with merging this. Not much to check, as most files are imported.

Let's wait till all dependencies are in so we can see if there's anything that needs fixing.

StefanJum reviewed (approved):
All works fine, will add the tag after the dependencies are met as @craciunoiuc suggested.

razvand commented:
@StefanJum , @mschlumpp , @craciunoiuc , this is in constant dependence of PR #566. Every time that changes, this needs updating. So let's just wait for that to get done. In the meanwhile, be sure to see if anything added (in the 5 commits of the PR) could be improved.

craciunoiuc reviewed (approved):
I'm not going over this again. I'll be an optimistic reviewer this time.

All good here. Thanks!

Reviewed-by: Cezar Craciunoiu <cezar.craciunoiu@unikraft.io>

jschlumpp reviewed (approved):
Approved-by: Marco Schlumpp <marco@unikraft.io>
